### ISSUE TITLE
Alternate way of detecting modified files

### ISSUE THREAD

**Issue by scribblemaniac** [morevnaproject/RenderChan#17]
Hey, I'm back! Well not quite yet, I still have a 2 more projects to finish before I can do more work on RenderChan, but soon!

So the problem is that some file system formats do not support file creation/modification dates, and even if they do the date is not always reliable. An example that affects all(?) systems is if you were to save a file without making any changes, then that file would needlessly be rerendered by RenderChan.

The solution I propose is to instead store a hash of the file (maybe in the .sync files?). Then when it comes time to check if the file has been updated, RenderChan hashes the file again compares it to the stored hash to see if any changes have been made. Python has builtin hashing algorithms in hashlib so this should be supported on all systems.

Another advantage of doing it this way is apparent if we make the hashing of the file something that modules could override. Then the module can remove parts that would not effect the render before hashing. For example, Inkscape could filter out comments and whitespace from SVGs before hashing them, so even if you change the comments or formatting of a file (which won't affect the rendered image), RenderChan won't need to rerender it.

**Comment by morevnaproject** [comment]
Hey, great to see you back! ^__^

This looks interesting as an option, but consider that with this algorithm the rendering process will get much slower (this will be especially notable when running RenderChan on the tree with no changes). This approach also will not be as effective as expected, see below.

> example that affects all(?) systems is if you were to save a file without making any changes

Well, most applications make "save" button inactive if you haven't made any changes. And there is issue with Blender: if you open file, then save (without making any change), then content of the file is still changed somehow. So, the md5sum will be different.

Still, I agree that timestamp might be not very reliable mechanism (especially on Windows). So, implementing this as an option would be a benefit.

**Comment by morevnaproject** [comment]
Also worth reading - http://scons.org/doc/0.98.4/HTML/scons-user/c779.html#AEN899

**Comment by scribblemaniac** [comment]
You're right, and I certainly wouldn't want to slow it down. I can implement this function in the same way as SCons "Decider function", relying first on the modification time and then checking the hash if the time is different. That should make the resulting speed decrease almost unnoticeable. Any problems with this?

**Comment by morevnaproject** [comment]
Looks good.  ^__^ Also, consider that we use timestamps for caching dependencies (render/cache.sqlite).
